### DISCUSSION TITLE
Question about the roadmap of parallel trie

### DISCUSSION THREAD

**Discussion by yutianwu** [paradigmxyz/reth#7982] [Technical discussions]
Hi, we researched `trie` and `parallel-trie` of `reth` recently, and wanna help with it. 

It seems parallel trie has been used in some places in the codebase like:

https://github.com/paradigmxyz/reth/blob/490fbb0f52e0b73eb65076912d42875093074694/crates/blockchain-tree/src/chain.rs#L221

But it has not been applied to the `Merkle Stage` so far. 

Can you share the concerns about the `parallel trie` and the road map about it?

**Comment by yutianwu** [comment]
@rkrasiuk hi, can you share some insights, thanks

**Comment by rkrasiuk** [comment]
hey @yutianwu, thanks for expressing the interest! we are not planning to integrate it into the merkle stage for now, but if we were, we would need to find a way to infer how many storage roots we should pre-calculate. merkle stage computes state root in iterations with incremental commits of intermediate nodes to the database. terminating condition for single iteration depends on how many entries we already walked during current iteration https://github.com/paradigmxyz/reth/blob/a590ed7ce5af3bbf6feb3cd5a626334311e341d7/crates/trie/src/trie.rs#L292-L314
